radixdlt-core/radixdlt/src/main/java/com/radixdlt/sync/LocalSyncService.java

```diff
@@ -0,0 +1,470 @@
+		this.syncCheckReceiveStatusTimeoutDispatcher.dispatch(
+			SyncCheckReceiveStatusTimeout.create(),
+			this.syncConfig.syncCheckReceiveStatusTimeout()
+		);
+	}
+
+	private ImmutableSet<BFTNode> choosePeersForSyncCheck() {
+		final var allPeers = this.addressBook.peers().collect(Collectors.toList());
+		Collections.shuffle(allPeers);
+		return allPeers.stream()
+			.limit(this.syncConfig.syncCheckMaxPeers())
+			.map(peer -> BFTNode.create(peer.getSystem().getKey()))
+			.collect(ImmutableSet.toImmutableSet());
+	}
+
+	public RemoteEventProcessor<StatusResponse> statusResponseEventProcessor() {
+		return this::processStatusResponse;
+	}
+
+	private void processStatusResponse(BFTNode peer, StatusResponse statusResponse) {
+		log.trace("LocalSync: Received status response {} from peer {}", statusResponse, peer);
+
+		if (!isSyncCheckState()) {
+			return;
+		}
+		final var syncCheckState = (SyncState.SyncCheckState) this.syncState;
+
+		if (!syncCheckState.hasAskedPeer(peer)) {
+			return; // we didn't ask this peer
+		}
+
+		if (syncCheckState.receivedResponseFrom(peer)) {
+			return; // already got the response from this peer
+		}
+
+		final var newState = syncCheckState.withStatusResponse(peer, statusResponse);
+		this.syncState = newState;
+
+		if (newState.gotAllResponses()) {
+			processPeerStatusResponsesAndStartSyncIfNeeded(newState); // we've got all the responses
+		}
+	}
+
+	private void processPeerStatusResponsesAndStartSyncIfNeeded(SyncState.SyncCheckState syncCheckState) {
+		// get the highest state that we received that is also higher than what we currently have
+		final var maybeMaxPeerHeader = syncCheckState.responses().values()
+			.stream()
+			.map(StatusResponse::getHeader)
+			.max(Comparator.comparing(VerifiedLedgerHeaderAndProof::getAccumulatorState, accComparator))
+			.filter(h ->
+				accComparator.compare(
+					h.getAccumulatorState(),
+					syncCheckState.getCurrentHeader().getAccumulatorState()
+				) > 0
+			);
+
+		maybeMaxPeerHeader.ifPresentOrElse(
+			maxPeerHeader -> {
+				// start sync with all peers that are at the highest received state
+				final var candidatePeers = syncCheckState.responses()
+					.entrySet().stream()
+					.filter(e ->
+						accComparator.compare(
+							e.getValue().getHeader().getAccumulatorState(),
+							maxPeerHeader.getAccumulatorState()
+						) == 0
+					)
+					.map(Map.Entry::getKey)
+					.collect(ImmutableList.toImmutableList());
+
+				this.startSync(candidatePeers, maxPeerHeader);
+			},
+			() -> {
+				// there is no peer ahead of us, retry the sync check after some delay
+				this.goToIdleAndScheduleSyncCheck();
+			});
+	}
+
+	public EventProcessor<SyncCheckReceiveStatusTimeout> syncCheckReceiveStatusTimeoutEventProcessor() {
+		return this::processSyncCheckReceiveStatusTimeout;
+	}
+
+	private void processSyncCheckReceiveStatusTimeout(SyncCheckReceiveStatusTimeout syncCheckReceiveStatusTimeout) {
+		if (!this.isSyncCheckState()) {
```

Comment:
Seems like pattern of `check state then ... do something` is a broken state machine pattern. Cleaner would be to have state itself process event and return next state.

Reply:
how about this (latest commit):
I made it a collection of functions (State, Event) -> State.
and then when processing an event it checks whether a transition exists given the event class and the current state.
The state variable is never directly accessed (or modified) inside a handler function.